### PR TITLE
Fix #127 - optimise index for sv queries

### LINKED ISSUE #127: Improve indexes
- Add right position to the index that already has left position (`coordinates`)
- Modify the indexes so they don't start with the same key (move the sv_type first to the index above)

### PR DESCRIPTION
**Pull request by dnil**
### This PR adds | fixes:
- Fix indexes according to our MongoDB consultant session conclusions

**How to prepare for test**:
- [ ] `ssh` to ...
- [ ] Install on stage:
`bash servers/resources/SERVER.scilifelab.se/update-[THIS_TOOL]-stage.sh [THIS-BRANCH-NAME]`

### How to test:
-

### Expected outcome:
- [ ]

### Review:
- [ ] Code approved by
- [ ] Tests executed by
- [ ] "Merge and deploy" approved by

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by dnil** [comment]
For any eventual index testing @northwestwitch!

**Comment by dnil** [comment]
This one should be safe, but will take some time to complete.

**Comment by dnil** [comment]
Testing on stage!

```
 [daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_loqusdb -t loqusdb -b optimise_index
Collecting git+https://github.com/Clinical-Genomics/loqusdb@optimise_index
  Cloning https://github.com/Clinical-Genomics/loqusdb (to revision optimise_index) to ./pip-req-build-9vpac9nt
  Running command git clone -q https://github.com/Clinical-Genomics/loqusdb /home/daniel.nilsson/tmp/pip-req-build-9vpac9nt
  Running command git checkout -b optimise_index --track origin/optimise_index
  Switched to a new branch 'optimise_index'
  Branch optimise_index set up to track remote branch optimise_index from origin.
  Resolved https://github.com/Clinical-Genomics/loqusdb to commit ceef096c9f45316b6f70dc5a751120f600f29350
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: loqusdb
  Building wheel for loqusdb (setup.py) ... done
  Created wheel for loqusdb: filename=loqusdb-2.7.1-py3-none-any.whl size=50210 sha256=7244409b2e9976073943ee064c5849dd8594aa269056cd5c489fb3af84f347a0
  Stored in directory: /home/daniel.nilsson/tmp/pip-ephem-wheel-cache-fu3wjsh0/wheels/1b/a5/43/10b672e63f2eb021adc7fb2d8fb85a2be5b57bc8def47b447e
Successfully built loqusdb
Installing collected packages: loqusdb
  Attempting uninstall: loqusdb
    Found existing installation: loqusdb 2.6.10
    Uninstalling loqusdb-2.6.10:
      Successfully uninstalled loqusdb-2.6.10
Successfully installed loqusdb-2.7.1
WARNING: You are using pip version 21.3.1; however, version 23.3.2 is available.
You should consider upgrading via the '/home/proj/stage/bin/miniconda3/envs/S_loqusdb/bin/python3.7 -m pip install --upgrade pip' command.
Collecting git+https://github.com/Clinical-Genomics/loqusdb@optimise_index
  Cloning https://github.com/Clinical-Genomics/loqusdb (to revision optimise_index) to ./pip-req-build-sdhuq7wo
  Running command git clone -q https://github.com/Clinical-Genomics/loqusdb /home/daniel.nilsson/tmp/pip-req-build-sdhuq7wo
  Running command git checkout -b optimise_index --track origin/optimise_index
  Switched to a new branch 'optimise_index'
  Branch optimise_index set up to track remote branch optimise_index from origin.
  Resolved https://github.com/Clinical-Genomics/loqusdb to commit ceef096c9f45316b6f70dc5a751120f600f29350
  Preparing metadata (setup.py) ... done
Requirement already satisfied: pytest==5.4.3 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (5.4.3)
Requirement already satisfied: cyvcf2==0.30.12 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (0.30.12)
Requirement already satisfied: mongomock==3.18.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (3.18.0)
Requirement already satisfied: click==7.1.2 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (7.1.2)
Requirement already satisfied: pymongo==3.7.1 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (3.7.1)
Requirement already satisfied: numpy==1.21.4 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (1.21.4)
Requirement already satisfied: coloredlogs==14.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (14.0)
Requirement already satisfied: pyyaml==5.4.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (5.4)
Requirement already satisfied: vcftoolbox==1.5 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (1.5)
Collecting pip==23.1.2
  Downloading pip-23.1.2-py3-none-any.whl (2.1 MB)
     |████████████████████████████████| 2.1 MB 12.1 MB/s
Collecting setuptools==65.5.1
  Using cached setuptools-65.5.1-py3-none-any.whl (1.2 MB)
Requirement already satisfied: mongo_adapter>=0.3.3 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (0.3.3)
Requirement already satisfied: ped_parser in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from loqusdb==2.7.1) (1.6.6)
Requirement already satisfied: humanfriendly>=7.1 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from coloredlogs==14.0->loqusdb==2.7.1) (10.0)
Requirement already satisfied: six in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from mongomock==3.18.0->loqusdb==2.7.1) (1.16.0)
Requirement already satisfied: sentinels in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from mongomock==3.18.0->loqusdb==2.7.1) (1.0.0)
Requirement already satisfied: attrs>=17.4.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (22.1.0)
Requirement already satisfied: wcwidth in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (0.2.5)
Requirement already satisfied: more-itertools>=4.0.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (8.14.0)
Requirement already satisfied: packaging in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (21.3)
Requirement already satisfied: pluggy<1.0,>=0.12 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (0.13.1)
Requirement already satisfied: py>=1.5.0 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (1.11.0)
Requirement already satisfied: importlib-metadata>=0.12 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from pytest==5.4.3->loqusdb==2.7.1) (4.12.0)
Requirement already satisfied: typing-extensions>=3.6.4 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from importlib-metadata>=0.12->pytest==5.4.3->loqusdb==2.7.1) (4.3.0)
Requirement already satisfied: zipp>=0.5 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from importlib-metadata>=0.12->pytest==5.4.3->loqusdb==2.7.1) (3.8.1)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/proj/stage/bin/miniconda3/envs/S_loqusdb/lib/python3.7/site-packages (from packaging->pytest==5.4.3->loqusdb==2.7.1) (3.0.9)
Installing collected packages: setuptools, pip
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.2.0
    Uninstalling setuptools-59.2.0:
      Successfully uninstalled setuptools-59.2.0
  Attempting uninstall: pip
    Found existing installation: pip 21.3.1
    Uninstalling pip-21.3.1:
      Successfully uninstalled pip-21.3.1
Successfully installed pip-23.1.2 setuptools-65.5.1
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/daniel.nilsson/tmp
done.
Log deploy... done.
loqusdb, version 2.7.1
θ64° [daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb index
2024-01-26 17:34:24 hasta.scilifelab.se loqusdb.commands.cli[188520] INFO Running loqusdb version 2.7.1
2024-01-26 17:34:24 hasta.scilifelab.se mongo_adapter.client[188520] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:34:24 hasta.scilifelab.se mongo_adapter.client[188520] INFO Connection established
2024-01-26 17:34:24 hasta.scilifelab.se mongo_adapter.adapter[188520] INFO Use database loqusdb-rd-stage
2024-01-26 17:34:24 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[188520] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:35:40 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[188520] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:35:40 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[188520] INFO creating indexes for collection identity: cluster, variant
```
Screenshot 2024-01-26 at 17 42 52

```
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-somatic index
2024-01-26 17:44:34 hasta.scilifelab.se loqusdb.commands.cli[200080] INFO Running loqusdb version 2.7.1
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.client[200080] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.client[200080] INFO Connection established
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.adapter[200080] INFO Use database loqusdb-somatic-stage
2024-01-26 17:44:34 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:44:35 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:44:35 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection identity: cluster, variant
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-tumor index
2024-01-26 17:44:40 hasta.scilifelab.se loqusdb.commands.cli[200396] INFO Running loqusdb version 2.7.1
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.client[200396] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.client[200396] INFO Connection established
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.adapter[200396] INFO Use database loqusdb-tumor-stage
2024-01-26 17:44:40 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:45:23 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:45:23 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection identity: cluster, variant
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-wes index
2024-01-26 17:45:39 hasta.scilifelab.se loqusdb.commands.cli[201516] INFO Running loqusdb version 2.7.1
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.client[201516] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.client[201516] INFO Connection established
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.adapter[201516] INFO Use database loqusdb-rd-wes-stage
2024-01-26 17:45:39 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:45:43 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:45:43 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection identity: cluster, variant
```

**Comment by dnil** [comment]
```
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-somatic index
2024-01-26 17:44:34 hasta.scilifelab.se loqusdb.commands.cli[200080] INFO Running loqusdb version 2.7.1
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.client[200080] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.client[200080] INFO Connection established
2024-01-26 17:44:34 hasta.scilifelab.se mongo_adapter.adapter[200080] INFO Use database loqusdb-somatic-stage
2024-01-26 17:44:34 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:44:35 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:44:35 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200080] INFO creating indexes for collection identity: cluster, variant
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-tumor index
2024-01-26 17:44:40 hasta.scilifelab.se loqusdb.commands.cli[200396] INFO Running loqusdb version 2.7.1
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.client[200396] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.client[200396] INFO Connection established
2024-01-26 17:44:40 hasta.scilifelab.se mongo_adapter.adapter[200396] INFO Use database loqusdb-tumor-stage
2024-01-26 17:44:40 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:45:23 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:45:23 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[200396] INFO creating indexes for collection identity: cluster, variant
[daniel.nilsson@hasta:~/tmp] [S_base] 5m56s $ loqusdb-wes index
2024-01-26 17:45:39 hasta.scilifelab.se loqusdb.commands.cli[201516] INFO Running loqusdb version 2.7.1
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.client[201516] INFO Connecting to uri:mongodb://loqusdb-stage:******@cg-mongo-stage.scilifelab.se:27030
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.client[201516] INFO Connection established
2024-01-26 17:45:39 hasta.scilifelab.se mongo_adapter.adapter[201516] INFO Use database loqusdb-rd-wes-stage
2024-01-26 17:45:39 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection variant: coordinates, end
2024-01-26 17:45:43 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection structural_variant: coordinates, short_coordinates
2024-01-26 17:45:43 hasta.scilifelab.se loqusdb.plugins.mongo.adapter[201516] INFO creating indexes for collection identity: cluster, variant
```

File: loqusdb/__init__.py [hunk 1]
```diff
@@ -28,10 +28,11 @@
     "structural_variant": [
         IndexModel(
             [
+                ("sv_type", ASCENDING),
                 ("chrom", ASCENDING),
                 ("end_chrom", ASCENDING),
-                ("sv_type", ASCENDING),
                 ("pos_left", ASCENDING),
+                ("pos_right", ASCENDING),
```
Comment (northwestwitch) [inline loqusdb/__init__.py:35]: 👍🏻

**Review by northwestwitch — APPROVED** [review]
Nice improvement!

**Comment by dnil** [comment]
Thanks for the review! I think we will wait with merging until after the OS update tomorrow in order not to complicate anything, should the index builds last longer than one day.

**Comment by northwestwitch** [comment]
> Thanks for the review! I think we will wait with merging until after the OS update tomorrow in order not to complicate anything, should the index builds last longer than one day.

Good plan!

**Comment by dnil** [comment]
Testing on stage, all four loqusdbs:
Screenshot 2024-03-18 at 10 33 59

Note for deployment to `prod`: let's not use the `loqusdb index` command, which drops all named indices. Instead, just do the one `structural_variant` collection index for all dbs.

**Comment by dnil** [comment]
Manually updated the `locusdb-*.structural_variant` coordinates index to start with `sv_type` and end with `pos_left` and lastly `pos_right`.

**Pull request merged**